src/tensorwaves/model/sympy.py

```diff
@@ -117,57 +119,45 @@ def optimized_lambdify(
         max_complexity=max_complexity,
     )
     if not sub_expressions:
-        return _backend_lambdify(top_expression, symbols, backend, **kwargs)
+        return lambdify(top_expression, symbols, backend, **kwargs)
 
     sorted_top_symbols = sorted(sub_expressions, key=lambda s: s.name)
-    top_function = _backend_lambdify(
+    top_function = lambdify(
         top_expression, sorted_top_symbols, backend, **kwargs
     )
-    sub_functions: List[Callable] = []
+    sub_functions: List[PositionalArgumentFunction] = []
     for symbol in tqdm(
         iterable=sorted_top_symbols,
         desc="Lambdifying sub-expressions",
         unit="expr",
         disable=not _use_progress_bar(),
     ):
         sub_expression = sub_expressions[symbol]
-        sub_function = _backend_lambdify(
-            sub_expression, symbols, backend, **kwargs
-        )
+        sub_function = lambdify(sub_expression, symbols, backend, **kwargs)
         sub_functions.append(sub_function)
 
     def recombined_function(*args: Any) -> Any:
-        new_args = [sub_function(*args) for sub_function in sub_functions]
-        return top_function(*new_args)
+        new_args = [func.function(*args) for func in sub_functions]
+        return top_function.function(*new_args)
```

Comment:
Ok I guess this answers my question above, but feels a bit like a design smell. Can't really follow the implementation details atm (not so up to date with the code details). So afaiu its about substituting parts of the function with sub functions and this takes care of the correct argument handling. Maybe you explain your thoughts

Reply:
https://github.com/ComPWA/tensorwaves/pull/353#discussion_r757888265
The next question though is whether those public frozen attributes should be part of the interface. I decided not to, because the sole function of `Function` is to compute an array from a `DataSample`.

But indeed, the implementation here is a bit wonky. It seems related to #323: ideal would be to offer a mechanism to lambdify an expression tree to a 'function tree' in terms of that back-end, where it's possible to inspect sub-functions. But the `Function` interface does not suffice there, because you need some attributes that tell which sub-functions a top function provides.

Reply:
https://github.com/ComPWA/tensorwaves/pull/353#issuecomment-982770331